gr-digital/lib/ofdm_cyclic_prefixer_impl.cc

```diff
@@ -31,135 +31,147 @@ namespace gr {
-		  io_signature::make (1, 1, sizeof(gr_complex)),
-		  len_tag_key),
-	d_fft_len(input_size),
-	d_output_size(output_size),
-	d_cp_size(output_size - input_size),
-	d_rolloff_len(rolloff_len),
-	d_up_flank((rolloff_len ? rolloff_len-1 : 0), 0),
-	d_down_flank((rolloff_len ? rolloff_len-1 : 0), 0),
-	d_delay_line(0, 0)
+    ofdm_cyclic_prefixer_impl::ofdm_cyclic_prefixer_impl(int fft_len, std::vector<int> cp_lengths, int rolloff_len, const std::string &len_tag_key)
+      : gr::tagged_stream_block
+        (
+          "ofdm_cyclic_prefixer",
+          gr::io_signature::make(1, 1, fft_len*sizeof(gr_complex)),
+          gr::io_signature::make(1, 1, sizeof(gr_complex)), len_tag_key
+        ),
+        d_fft_len(fft_len),
+        d_state(0),
+        d_cp_max(0),
+        d_rolloff_len(rolloff_len),
+        d_cp_lengths(cp_lengths),
+        d_up_flank((rolloff_len ? rolloff_len-1 : 0), 0),
+        d_down_flank((rolloff_len ? rolloff_len-1 : 0), 0),
+        d_delay_line(0, 0),
+        d_len_tag_key(len_tag_key)
     {
-      set_relative_rate(d_output_size);
-
-      // Flank of length 1 would just be rectangular
-      if (d_rolloff_len == 1) {
-	d_rolloff_len = 0;
+      for(const int& cp_len : d_cp_lengths) {
+        d_cp_max = std::max(d_cp_max, cp_len);
       }
-      if (d_rolloff_len) {
-	d_delay_line.resize(d_rolloff_len-1, 0);
-	if (rolloff_len > d_cp_size) {
-	  throw std::invalid_argument("cyclic prefixer: rolloff len must smaller than the cyclic prefix.");
-	}
-	// The actual flanks are one sample shorter than d_rolloff_len, because the
-	// first sample of the up- and down flank is always zero and one, respectively
-	for (int i = 1; i < d_rolloff_len; i++) {
-	  d_up_flank[i-1]   = 0.5 * (1 + cos(M_PI * i/rolloff_len - M_PI));
-	  d_down_flank[i-1] = 0.5 * (1 + cos(M_PI * (rolloff_len-i)/rolloff_len - M_PI));
-	}
+      // Sanity
+      for(unsigned i=0; i<d_cp_lengths.size(); i++) {
+        if (d_cp_lengths[i] != 0) {
+          break;
+        }
+        if (i == d_cp_lengths.size()-1) {
+          throw std::invalid_argument(this->alias() + std::string(": It does not make sense to only have one or a couple of CPs with length 0."));
+        }
       }
-
-      if (len_tag_key.empty()) {
-	set_output_multiple(d_output_size);
-      } else {
-	set_tag_propagation_policy(TPP_DONT);
+      if (d_cp_lengths.empty()) {
+        throw std::invalid_argument(this->alias() + std::string(": CP lengths vector can not be empty. Please enter at least (d,) (a single Python tuple) with d being a integer CP length."));
+      }
+      // Give the buffer allcoator and scheduler a hint about the ratio between input and output.
+      set_relative_rate(d_cp_max + d_fft_len);
+      // Flank of length 1 would just be rectangular.
+      if(d_rolloff_len == 1) {
+        throw std::invalid_argument(this->alias() + std::string(": A rolloff length of 1 does not make sense, as it results in a boxcar function."));
```

Comment:
This is another API change. I think I made the previous version because there are cases when people think they need to provide 1 when what they mean is 0. I would strongly advocate leaving the old behaviour unchanged, but add a warning-level log message.